### PR TITLE
Tfrt update

### PR DESCRIPTION
**Pull request by alexwitt2399**
Draft PR to start the review process on the large number of updates and changes. There are still a couple things to do: 

- finish `optimize_od.py`.  **DONE** https://github.com/uavaustin/target-finder-model/pull/6/commits/45269ef130b16e380fede55f0ca4301d1dcd5406

- complete the Github Actions CI with a slack notifications.  **DONE**

Whenever this branch is merged, I’ll squash a lot of the CI commits. `TODO`

- make sure tftrt can be used on the Xavier in the `:xavier` container.
  - this will be a good chance for the members to get the docker image setup on the xavier and test the scripts

There  are a lot of code changes here. Feel free to take a week or so to review them, but I will do the merge!

### CODE REVIEW

**Comment by alexwitt2399** [comment]
`optimize_od.py` works now after building pycocotools from scratch. Appears that TensorFlow request a version that conflicts with the numpy version.

**Comment by alexwitt2399** [comment]
complete the Github Actions CI with a slack notifications. **DONE**. https://github.com/uavaustin/target-finder-model/pull/6/commits/45269ef130b16e380fede55f0ca4301d1dcd5406

File: scripts_generate/create_detection_data.py [hunk 1]
```diff
@@ -247,21 +308,21 @@ def _strip_image(image):
 
 def _add_alphanumeric(image, shape, alpha, alpha_rgb, font_file):
     # Adjust alphanumeric size based on the shape it will be on
-    if shape == 'star':
+    if shape == "star":
```
Comment (oldmud0) [inline scripts_generate/create_detection_data.py]: I see you have converted

File: .coveralls.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+service_name: github-actions
+repo_token: 4K7LtmFlYnTWaMYypWe8fy6ruMnAfT1UG
```
Comment (sshh12) [inline .coveralls.yml]: Is this ok to be public?

File: README.md [hunk 3]
```diff
@@ -1,60 +1,118 @@
...
+```
+### Optimize Object Detector Model
+```
+python scripts_tf/optimize_od.py \
+    --input_saved_model_dir=models/MODEL_NAME/saved_model \
+    --output_saved_model_dir=models/MODEL_NAME/optimized \
+    --data_dir model_data/records/ \
+    --use_trt \
+    --precision FP32
+```
```
Comment (sshh12) [inline README.md]: In the future it prob be helpful to just build a cli do automate these tedious cmds. Good for now thou.

File: README.md [hunk 4]
```diff
@@ -1,60 +1,118 @@
...
+#### Model Files
+* `model_data/` Object detection training data, filled w/generate scripts
+* `model/` Model configs, TensorFlow pipelines, training checkpoints
+
+#### Scripts
+* `scripts_generate/` Scripts for generating artifical training data
+* `scripts_tf/` Helpful TensorFlow utilities
+
+#### Misc
+* `.circleci/` CI Config for automated builds and testing
```
Comment (sshh12) [inline README.md]: Are we still using circleci?

File: README.md [hunk 5]
```diff
@@ -1,60 +1,118 @@
...
+* `Dockerfiles/` Docker resources for creating a prebuilt ML environment
+* `target_finder_model/` The package that will be exported to [target-finder](https://github.com/uavaustin/target-finder) when a release is created.
+
+## Testing
+Tests can be executed by running `make test` inside `Dockerfiles`.
 
-Full builds are run on tags and the model is uploaded as a build artifact at
-the end and pushed to GitHub Releases.
+## Readability 
+All Python code conforms to the [Google Python Style Guide](http://google.github.io/styleguide/pyguide.html). The code can be formatted using [Black](https://black.readthedocs.io/en/stable/).
```
Comment (sshh12) [inline README.md]: **Bold**

File: experimental/entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+echo "Creating TF-Records."
+python scripts_tf/create_tf_records.py
+
+cp -r model_data/records /host/model_data
+
+echo "Creating label map."
+python scripts_tf/create_label_map.py
+
+echo "Training classification model."
+#python scripts_tf/train_clf.py
```
Comment (sshh12) [inline experimental/entrypoint.sh]: Should this be commented out?

File: experimental/entrypoint.sh [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+# This command will fail on CI builds without gpu ()
+if [ $? -eq 0 ]; then
```
Comment (sshh12) [inline experimental/entrypoint.sh]: It might be better to do an `if isCI...else...` rather than ignoring failed output.

File: scripts_generate/build.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+
+if __name__ == "__main__":
+    # Try to get env vars from tox build
+    try:
+        NUM_IMAGES = int(os.environ["NUM_IMAGES"])
+        NUM_VAL_IMAGES = int(os.environ["NUM_VAL_IMAGES"])
+    except Exception:
+        NUM_IMAGES = 20
+        NUM_VAL_IMAGES = 10
```
Comment (sshh12) [inline scripts_generate/build.py]: I think one of these should be from the config yaml.

File: scripts_generate/create_clf_data.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        image_names,
+        gen_types,
+    )
+
+    with multiprocessing.Pool(None) as pool:
+        processes = pool.imap_unordered(_single_clf_image, data)
+        for i in tqdm(processes, total=num_gen):
+            pass
+
+    return
```
Comment (sshh12) [inline scripts_generate/create_clf_data.py]: I think this can be removed.

File: scripts_generate/create_clf_data.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    )
+    background.save(bkg_fn)
+
+    # Now consider the shape image
+    shape = Image.open(shape_img).resize(config.PRECLF_SIZE)
+    shape = _enhance_image(shape, enhancement)
+    shape_fn = os.path.join(data_path, "target_{}.{}".format(number, config.IMAGE_EXT))
+    shape.save(shape_fn)
+
+    return
```
Comment (sshh12) [inline scripts_generate/create_clf_data.py]: Another extra return.

File: scripts_generate/create_detection_data.py [hunk 11]
```diff
@@ -50,52 +53,78 @@ def generate_all_shapes(gen_type, num_gen, offset=0):
...
+                    angles,
+                    target_colors,
+                    target_rgbs,
+                    alpha_colors,
+                    alpha_rgbs,
+                    xs,
+                    ys,
+                )
+            )
+        )
```
Comment (sshh12) [inline scripts_generate/create_detection_data.py]: lol lisp style

File: scripts_generate/create_detection_data.py [hunk 12]
```diff
@@ -247,21 +308,21 @@ def _strip_image(image):
 
 def _add_alphanumeric(image, shape, alpha, alpha_rgb, font_file):
     # Adjust alphanumeric size based on the shape it will be on
-    if shape == 'star':
+    if shape == "star":
         font_multiplier = 0.14
-    if shape == 'triangle':
+    if shape == "triangle":
```
Comment (sshh12) [inline scripts_generate/create_detection_data.py]: In the future this can be prob turned into a map (shape->shape_consts).

File: scripts_tf/create_label_map.py [hunk 13]
```diff
@@ -0,0 +1,22 @@
+"""
+Create label map file.
+"""
+import os
+
+with open(
+    os.path.join(os.path.dirname(__file__), os.pardir, "config.yaml"), "r"
+) as stream:
```
Comment (sshh12) [inline scripts_tf/create_label_map.py]: looks kinda weird imo, but it works

File: target_finder_model/inference.py [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    def predict(self, input_data, batch_size=4):
+
+        if len(input_data) == 0:
+            return []
+        elif isinstance(input_data, list):
+            # allow list of paths as input
+            input_data = np.array(
+                [np.asarray(fn) for fn in input_data], dtype=np.uint8)
+        else:
+            pass
```
Comment (sshh12) [inline target_finder_model/inference.py]: Delete

File: .coveralls.yml (hunk 2 above)
Comment (oldmud0) [reply .coveralls.yml]: Nope. If we are still sending this through Travis CI, the `repo_token` entry can be removed entirely. Otherwise, the Coveralls action on GH Actions should be used.

File: scripts_generate/create_detection_data.py (hunk 11 above)
Comment (oldmud0) [reply scripts_generate/create_detection_data.py]: This is why I dislike black sometimes.

File: .coveralls.yml (hunk 2 above)
Comment (alexwitt2399) [reply .coveralls.yml]: I will look into the action. FYI, the public tokens are in the orchestra CI's, (link)[https://github.com/uavaustin/orchestra/blob/e30965fd71b0f9e4de82d0f8a5cef6770b60a4db/.travis.yml#L45]

Comment (oldmud0) [reply .coveralls.yml]: Those tokens in the `.travis.yml` file are encrypted; when Travis CI performs the build, it decrypts the tokens based on a secret token only known within the build environment.

File: README.md (hunk 4 above)
Comment (alexwitt2399) [reply README.md]: No, I'll remove that.

File: .coveralls.yml (hunk 2 above)
Comment (alexwitt2399) [reply .coveralls.yml]: Got it! Thanks!

File: README.md (hunk 3 above)
Comment (alexwitt2399) [reply README.md]: Agreed. I'll make a issue on the projects for this semester.

File: experimental/entrypoint.sh (hunk 6 above)
Comment (alexwitt2399) [reply experimental/entrypoint.sh]: Not exactly, I'll add a blurb on the readme about experimental code

File: scripts_generate/create_clf_data.py (hunk 10 above)
Comment (alexwitt2399) [reply scripts_generate/create_clf_data.py]: Fixed

File: scripts_generate/create_detection_data.py (hunk 11 above)
Comment (alexwitt2399) [reply scripts_generate/create_detection_data.py]: Yeah it is a little awkward

File: scripts_generate/create_detection_data.py (hunk 12 above)
Comment (alexwitt2399) [reply scripts_generate/create_detection_data.py]: Good idea.

**Comment by alexwitt2399** [comment]
I will get back with you on model inference times and accuracy metrics @oldmud0. 

I need to tweak the models for the Xavier, then I'll have a better idea. 

Looking at the results so far, I'd expect to get about 1-3 sec per image. I might change the preclf to a smaller model.

File: target_finder_model/inference.py (hunk 14 above)
Comment (alexwitt2399) [reply target_finder_model/inference.py]: Done

File: scripts_tf/create_label_map.py (hunk 13 above)
Comment (alexwitt2399) [reply scripts_tf/create_label_map.py]: Yeah, it's a bit awkward. Let's run with black for a little and if it's too cumbersome we can toss it.

**Comment by alexwitt2399** [comment]
@oldmud0 When it comes to model timing on the Xavier, I am not sure yet because I need to get an hdmi cable so I can run things on my xavier here.

But here are the model metrics as of now (batch size of 1):

Classification: 
```Classified 2000 images in 38.722596406936646 s. Average 51.649429159707026 imgs/sec
Classification accuracy: 0.9525
```

Detection:
```
1999 images in 87.49793457984924 s. Average 22.846253566999845 imgs/s
{'TP': 1956, 'FP': 91, 'FN': 2044}
Recall: 0.489
Precision: 0.9555446995603322
```

The low recall due to the model struggling to find alphas

File: scripts_generate/create_clf_data.py (hunk 9 above)
Comment (alexwitt2399) [reply scripts_generate/create_clf_data.py]: Removed

File: scripts_generate/build.py (hunk 8 above)
Comment (alexwitt2399) [reply scripts_generate/build.py]: Good catch

File: experimental/entrypoint.sh (hunk 7 above)
Comment (alexwitt2399) [reply experimental/entrypoint.sh]: Yeah this script could use some work. I am going to leave it in experimental since its WIP, but maybe could be tweaked for future use.

File: .coveralls.yml (hunk 2 above)
Comment (alexwitt2399) [reply .coveralls.yml]: I think I will just take out the coveralls for now since the unit tests on tfm are really minimal. Perhaps could add more comprehensive tests later..

**Comment by alexwitt2399** [comment]
Squashing done. CI done. Merging
